example/glue/EnvironmentDelegate.hpp

```diff
@@ -0,0 +1,245 @@
+/*******************************************************************************
+ *
+ * (c) Copyright IBM Corp. 2017
+ *
+ *  This program and the accompanying materials are made available
+ *  under the terms of the Eclipse Public License v1.0 and
+ *  Apache License v2.0 which accompanies this distribution.
+ *
+ *      The Eclipse Public License is available at
+ *      http://www.eclipse.org/legal/epl-v10.html
+ *
+ *      The Apache License v2.0 is available at
+ *      http://www.opensource.org/licenses/apache2.0.php
+ *
+ * Contributors:
+ *    Multiple authors (IBM Corp.) - initial implementation and documentation
+ ******************************************************************************/
+
+#ifndef ENVIRONMENTDELEGATE_HPP_
+#define ENVIRONMENTDELEGATE_HPP_
+
+class MM_EnvironmentBase;
+
+/**
+ * The GC_Environment class is opaque to OMR and may be used by the client language to
+ * maintain language-specific information relating to a OMR VM thread, for example, local
+ * buffers or caches that are maintained on a per thread basis.
+ */
+class GC_Environment
+{
+	/* Data members */
+private:
+
+protected:
+
+public:
+
+	/* Function members */
+private:
+
+protected:
+
+public:
+	GC_Environment() {}
+};
+
+/***
+ * NOTE: The OMR VM access model grants exclusive VM access to one thread
+ * only if no other thread has VM access. Conversely one or more threads may
+ * be granted shared VM access only if no other thread holds exclusive
+ * VM access. This can be modeled as a write-bias reader/writer lock where
+ * threads holding shared VM access periodically check for pending
+ * requests for exclusive VM access and relinquish VM access in that event.
+ *
+ * It is assumed that a thread owning the exclusive lock may attempt to
+ * reacquire exclusive access when it already has exclusive access. Each
+ * OMR thread maintains a counter to track the number of times it has acquired
+ * the lock and not released it. This counter (OMR_VMThread::exclusiveCount)
+ * must be maintained in the language-specific environment language interface
+ * (this file).
+ *
+ * OMR threads must acquire shared VM access and hold it as long as they are
+ * accessing VM internal structures. In the event that another thread requests
+ * exclusive VM access, threads holding shared VM access must release shared
+ * VM access and suspend activities involving direct access to VM structures.
+ *
+ * This (example) environment delegate implements a simplistic VM access framework
+ * that is not pre-emptive; that is, it relies on threads that are holding non-
+ * exclusive VM access to check frequently (sub-millisecond) whether any other
```

Comment:
Yup. What is really needed here is a preemptive framework that blocks mutator threads at safe points when an exclusive access request (for GC) is pending. Since we do not yet have a generic framework for preemption in omr, mutator threads must poll to check for this condition and, if the polling frequency is lower than a few ms, GC cycles will be held up longer than they should be.

So sub-millisecond is just a guideline and suggests some urgency around this, as is appropriate.  A lot can happen in a millisecond, and polling once or twice in that interval would introduce less slowdown than would a thread that did not poll and yield for >1 ms while a request for exclusive VM access was pending.